File: backend/app/services/flashscore_watch.py
from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import Settings, get_settings
from app.models import StatisticalConfig
from app.schemas.api import (
    FlashscoreGoalEmailRequest,
    FlashscoreMatchRead,
    FlashscoreTickResult,
    FlashscoreWatchState,
    SofaScoreGoalIncident,
    SofaScoreTeamEvent,
)
from app.services.email_alerts import send_flashscore_goal_email
from app.services.sofascore_crawlora_provider import fetch_event_incidents, fetch_live_events
from app.utils.team_match import same_team


FLASHSCORE_WATCH_KEY = "flashscore_watch"
ALERT_ODDS_THRESHOLD = 1.5
EARLY_GOAL_MINUTE = 30


def save_flashscore_watch(
    db: Session,
    *,
    day: int,
    captured_at: datetime | None,
    matches: list[FlashscoreMatchRead],
) -> FlashscoreWatchState:
    stamp = captured_at or datetime.now(UTC)
    payload = {
        "day": day,
        "captured_at": stamp.isoformat(),
        "matches": [match.model_dump(mode="json") for match in matches],
        "updated_at": datetime.now(UTC).isoformat(),
    }
    config = db.scalar(select(StatisticalConfig).where(StatisticalConfig.key == FLASHSCORE_WATCH_KEY))
    if not config:
        config = StatisticalConfig(
            key=FLASHSCORE_WATCH_KEY,
            description="Flashscore low-odds watchlist for SofaScore signal ticks.",
            value=payload,
        )
        db.add(config)
    else:
        config.value = payload
        config.description = "Flashscore low-odds watchlist for SofaScore signal ticks."
    db.commit()
    db.refresh(config)
    return _watch_from_payload(config.value)


def load_flashscore_watch(db: Session) -> FlashscoreWatchState | None:
    config = db.scalar(select(StatisticalConfig).where(StatisticalConfig.key == FLASHSCORE_WATCH_KEY))
    if not config or not isinstance(config.value, dict):
        return None
    return _watch_from_payload(config.value)


def clear_flashscore_watch(db: Session) -> None:
    config = db.scalar(select(StatisticalConfig).where(StatisticalConfig.key == FLASHSCORE_WATCH_KEY))
    if not config:
        return
    config.value = {
        "day": 0,
        "captured_at": None,
        "matches": [],
        "updated_at": datetime.now(UTC).isoformat(),
    }
    db.commit()


def merge_flashscore_with_sofascore(
    matches: list[FlashscoreMatchRead],
    events: list[SofaScoreTeamEvent],
) -> list[FlashscoreMatchRead]:
    merged: list[FlashscoreMatchRead] = []
    for match in matches:
        event = next(
            (
                candidate
                for candidate in events
                if same_team(match.home_team, candidate.home_team)
                and same_team(match.away_team, candidate.away_team)
            ),
            None,
        )
        data = match.model_dump()
        if event is not None:
            data["status"] = event.status or match.status
            # Always take the freshly computed live minute; never keep a stale captured value
            # when the provider has live timing for the event.
            if event.minute is not None:
                data["minute"] = event.minute
                data["minute_extra"] = event.minute_extra
            data["home_score"] = event.home_score if event.home_score is not None else match.home_score
            data["away_score"] = event.away_score if event.away_score is not None else match.away_score
            data["sofascore_event_id"] = event.event_id
        merged.append(with_early_goal_flags(FlashscoreMatchRead.model_validate(data)))
    return merged


def apply_goal_incidents(
    match: FlashscoreMatchRead,
    goals: list[SofaScoreGoalIncident],
) -> FlashscoreMatchRead:
    """Attach the real goal minutes from the SofaScore timeline and re-evaluate flags.

    An empty timeline (e.g. a transient provider glitch) never wipes a minute already
    detected — the previously stored minutes are kept.
    """
    if not goals:
        return with_early_goal_flags(match)
    home_minutes = sorted({goal.minute for goal in goals if goal.is_home})
    away_minutes = sorted({goal.minute for goal in goals if not goal.is_home})
    data = match.model_dump()
    data["home_goal_minutes"] = home_minutes
    data["away_goal_minutes"] = away_minutes
    return with_early_goal_flags(FlashscoreMatchRead.model_validate(data))


def favorite_early_goal_minute(match: FlashscoreMatchRead) -> int | None:
    """First real minute (≤30) the favorite scored, when the timeline is known."""
    if match.favorite_side not in {"home", "away"}:
        return None
    minutes = match.away_goal_minutes if match.favorite_side == "away" else match.home_goal_minutes
    early = [minute for minute in minutes if minute <= EARLY_GOAL_MINUTE]
    return min(early) if early else None


def with_early_goal_flags(match: FlashscoreMatchRead) -> FlashscoreMatchRead:
    minute = match.minute
    home_score = match.home_score or 0
    away_score = match.away_score or 0
    total_goals = home_score + away_score
    favorite_score = away_score if match.favorite_side == "away" else home_score
    in_early_window = minute is not None and minute <= EARLY_GOAL_MINUTE

    # Real goal minutes from the SofaScore timeline take priority over the poll-minute guess.
    home_minutes = sorted(match.home_goal_minutes or [])
    away_minutes = sorted(match.away_goal_minutes or [])
    favorite_minutes = away_minutes if match.favorite_side == "away" else home_minutes
    early_incident_minutes = [value for value in home_minutes + away_minutes if value <= EARLY_GOAL_MINUTE]
    favorite_early_incident = [value for value in favorite_minutes if value <= EARLY_GOAL_MINUTE]
    has_incidents = bool(home_minutes or away_minutes)

    favorite_watched = (
        match.favorite_team is not None
        and match.favorite_odds is not None
        and match.favorite_odds <= ALERT_ODDS_THRESHOLD
    )

    if has_incidents:
        saw_early_goal = bool(match.early_goal) or bool(early_incident_minutes)
        saw_early_favorite_goal = bool(match.early_favorite_goal) or (
            favorite_watched and bool(favorite_early_incident)
        )
    else:
        saw_early_goal = bool(match.early_goal) or (in_early_window and total_goals > 0)
        saw_early_favorite_goal = bool(match.early_favorite_goal) or (
            in_early_window and favorite_watched and favorite_score > 0
        )

    # Prefer the true incident minute; fall back to the previously stored or current poll minute.
    if early_incident_minutes:
        early_goal_minute = min(early_incident_minutes)
    elif match.early_goal_minute is not None:
        early_goal_minute = match.early_goal_minute
    elif saw_early_goal and in_early_window:
        early_goal_minute = minute
    else:
        early_goal_minute = None

    # First goal of the match (any team), taken only from the timeline and kept sticky
    # once detected. Never inferred from the current scoreline.
    all_incident_minutes = sorted(home_minutes + away_minutes)
    first_goal_minute = all_incident_minutes[0] if all_incident_minutes else match.first_goal_minute
    goal_under_30 = first_goal_minute is not None and first_goal_minute <= EARLY_GOAL_MINUTE

    return match.model_copy(
        update={
            "home_goal_minutes": home_minutes,
            "away_goal_minutes": away_minutes,
            "early_goal": saw_early_goal,
            "early_favorite_goal": saw_early_favorite_goal,
            "early_goal_minute": early_goal_minute,
            "first_goal_minute": first_goal_minute,
            "goal_under_30": goal_under_30,
            "alert_eligible": saw_early_favorite_goal or is_alert_eligible(match),
        }
    )


def is_alert_eligible(match: FlashscoreMatchRead) -> bool:
    if match.early_favorite_goal:
        return True
    if (
        not match.favorite_team
        or match.favorite_odds is None
        or match.favorite_odds > ALERT_ODDS_THRESHOLD
        or match.minute is None
        or match.minute > EARLY_GOAL_MINUTE
        or match.home_score is None
        or match.away_score is None
    ):
        return False
    favorite_score = match.away_score if match.favorite_side == "away" else match.home_score
    return (favorite_score or 0) > 0


def needs_live_poll(match: FlashscoreMatchRead, now: datetime | None = None) -> bool:
    """True when a ≤1.50 favorite can still produce a timely early-goal signal."""
    if match.favorite_odds is None or match.favorite_odds > ALERT_ODDS_THRESHOLD:
        return False
    if match.early_favorite_goal or match.alert_eligible:
        return False
    status = (match.status or "").lower()
    if "finish" in status or "ended" in status or "afterpen" in status:
        return False
    if match.minute is not None:
        return match.minute <= EARLY_GOAL_MINUTE + 10
    if match.start_time is None:
        return True
    current = now or datetime.now(UTC)
    start = match.start_time
    if start.tzinfo is None:
        start = start.replace(tzinfo=UTC)
    return start - timedelta(minutes=20) <= current <= start + timedelta(minutes=50)


def _enrich_eligible_with_incidents(match: FlashscoreMatchRead) -> FlashscoreMatchRead:
    """Fetch the SofaScore goal timeline for an alert-eligible match, if not already known.

    Bounded to alert-eligible matches to keep Crawlora credit usage proportional to signals.
    """
    if not is_alert_eligible(match) or match.sofascore_event_id is None:
        return match
    if match.home_goal_minutes or match.away_goal_minutes:
        return match
    try:
        incidents = fetch_event_incidents(match.sofascore_event_id)
    except Exception:  # noqa: BLE001 - never break the tick on a timeline fetch
        return match
    return apply_goal_incidents(match, incidents.goals)


def run_flashscore_signal_tick(db: Session, settings: Settings | None = None) -> FlashscoreTickResult:
    settings = settings or get_settings()
    watch = load_flashscore_watch(db)
    if watch is None or not watch.matches:
        return FlashscoreTickResult(
            status="no_watch",
            checked=0,
            eligible=0,
            emails_sent=0,
            message="No hay lista vigilada. Captura cuotas ≤ 1,60 en la web para activar señales en segundo plano.",
        )

    if not settings.crawlora_api_key:
        return FlashscoreTickResult(
            status="provider_not_configured",
            checked=len(watch.matches),
            eligible=0,
            emails_sent=0,
            message="Falta CRAWLORA_API_KEY para revisar goles con SofaScore.",
        )

    critical = [match for match in watch.matches if needs_live_poll(match)]
    if not critical:
        return FlashscoreTickResult(
            status="idle",
            checked=len(watch.matches),
            eligible=0,
            emails_sent=0,
            message=(
                f"{len(watch.matches)} vigilados, ninguno en ventana critica. "
                "Sin consulta SofaScore (ahorra creditos Crawlora)."
            ),
        )

    try:
        live = fetch_live_events("football")
    except Exception as exc:  # noqa: BLE001 - surface provider failures to the tick caller
        return FlashscoreTickResult(
            status="request_failed",
            checked=len(watch.matches),
            eligible=0,
            emails_sent=0,
            message=f"No se pudo consultar SofaScore live: {exc}",
        )

    merged = merge_flashscore_with_sofascore(watch.matches, live.events)
    # Enrich alert-eligible matches with the real goal timeline so the email reports the true minute.
    merged = [_enrich_eligible_with_incidents(match) for match in merged]
    save_flashscore_watch(
        db,
        day=watch.day,
        captured_at=watch.captured_at,
        matches=merged,
    )

    eligible = [match for match in merged if is_alert_eligible(match)]
    emails_sent = 0
    for match in eligible:
        if (
            not match.favorite_team
            or match.favorite_odds is None
            or match.minute is None
            or match.home_score is None
            or match.away_score is None
        ):
            continue
        result = send_flashscore_goal_email(
            FlashscoreGoalEmailRequest(
                event_id=match.event_id,
                competition=match.competition,
                home_team=match.home_team,
                away_team=match.away_team,
                favorite_team=match.favorite_team,
                favorite_odds=match.favorite_odds,
                minute=favorite_early_goal_minute(match) or match.minute,
                home_score=match.home_score,
                away_score=match.away_score,
            ),
            settings,
        )
        if result.status in {"sent", "deduplicated"} and result.sent:
            if result.status == "sent":
                emails_sent += 1

    return FlashscoreTickResult(
        status="ok",
        checked=len(merged),
        eligible=len(eligible),
        emails_sent=emails_sent,
        message=(
            f"SofaScore tick · {len(merged)} vigilados · {len(live.events)} live · "
            f"{len(eligible)} señales · {emails_sent} emails nuevos."
        ),
    )


def _watch_from_payload(payload: dict[str, Any]) -> FlashscoreWatchState:
    matches_raw = payload.get("matches") or []
    matches = [FlashscoreMatchRead.model_validate(item) for item in matches_raw]
    captured_raw = payload.get("captured_at")
    captured_at = None
    if isinstance(captured_raw, str) and captured_raw:
        captured_at = datetime.fromisoformat(captured_raw.replace("Z", "+00:00"))
    updated_raw = payload.get("updated_at")
    updated_at = None
    if isinstance(updated_raw, str) and updated_raw:
        updated_at = datetime.fromisoformat(updated_raw.replace("Z", "+00:00"))
    return FlashscoreWatchState(
        day=int(payload.get("day") or 0),
        captured_at=captured_at,
        updated_at=updated_at,
        matches=[with_early_goal_flags(match) for match in matches],
    )

File: frontend/src/utils/flashscoreWatch.ts
import type { FlashscoreMatch, SofaScoreGoalIncident, SofaScoreTeamEvent } from "../types/api";
import { sameTeam } from "./teamMatch";

export const FLASHSCORE_WATCH_KEY = "within_flashscore_watch_v1";
export const ALERT_ODDS_THRESHOLD = 1.5;
export const LIST_ODDS_THRESHOLD = 1.6;
export const EARLY_GOAL_MINUTE = 30;
/** Poll SofaScore every minute while alert candidates are in the early window. */
export const FAST_LIVE_REFRESH_MS = 60 * 1000;
export const SLOW_LIVE_REFRESH_MS = 5 * 60 * 1000;

export type FlashscoreWatchState = {
  capturedAt: string;
  day: number;
  matches: FlashscoreMatch[];
};

export function readFlashscoreWatch(): FlashscoreWatchState | null {
  try {
    const raw = JSON.parse(localStorage.getItem(FLASHSCORE_WATCH_KEY) ?? "null");
    if (!raw || typeof raw !== "object" || !Array.isArray(raw.matches)) {
      return null;
    }
    return {
      capturedAt: typeof raw.capturedAt === "string" ? raw.capturedAt : new Date().toISOString(),
      day: typeof raw.day === "number" ? raw.day : 0,
      matches: raw.matches.filter(isFlashscoreMatch).map(withEarlyGoalFlags),
    };
  } catch {
    return null;
  }
}

export function writeFlashscoreWatch(state: FlashscoreWatchState) {
  localStorage.setItem(FLASHSCORE_WATCH_KEY, JSON.stringify(state));
}

export function clearFlashscoreWatch() {
  localStorage.removeItem(FLASHSCORE_WATCH_KEY);
}

export function mergeFlashscoreWithSofaScore(
  matches: FlashscoreMatch[],
  events: SofaScoreTeamEvent[],
): FlashscoreMatch[] {
  return matches.map((match) => {
    const event = events.find((candidate) =>
      sameTeam(match.home_team, candidate.home_team) &&
      sameTeam(match.away_team, candidate.away_team)
    );
    const base = event
      ? {
          ...match,
          status: event.status || match.status,
          // Always take the freshly computed live minute; never keep a stale captured value
          // when the provider has live timing for the event.
          minute: event.minute != null ? event.minute : match.minute,
          minute_extra: event.minute != null ? event.minute_extra ?? null : match.minute_extra,
          home_score: event.home_score ?? match.home_score,
          away_score: event.away_score ?? match.away_score,
          sofascore_event_id: event.event_id,
        }
      : { ...match };
    return withEarlyGoalFlags(base);
  });
}

/**
 * Attach the real goal minutes from the SofaScore timeline and re-evaluate flags.
 * An empty timeline never wipes a minute already detected.
 */
export function applyGoalIncidents(match: FlashscoreMatch, goals: SofaScoreGoalIncident[]): FlashscoreMatch {
  if (goals.length === 0) {
    return withEarlyGoalFlags(match);
  }
  const homeMinutes = uniqueSortedMinutes(goals.filter((goal) => goal.is_home).map((goal) => goal.minute));
  const awayMinutes = uniqueSortedMinutes(goals.filter((goal) => !goal.is_home).map((goal) => goal.minute));
  return withEarlyGoalFlags({
    ...match,
    home_goal_minutes: homeMinutes,
    away_goal_minutes: awayMinutes,
  });
}

/**
 * Whether the SofaScore goal timeline should be (re)fetched for this match.
 * Keeps retrying while the first goal minute has not been captured yet (the timeline
 * often lags a few seconds behind the score), and refetches when a new goal arrives.
 */
export function shouldFetchIncidents(match: FlashscoreMatch, coveredTotal = 0): boolean {
  if (match.sofascore_event_id == null) {
    return false;
  }
  const total = (match.home_score ?? 0) + (match.away_score ?? 0);
  if (total <= 0) {
    return false;
  }
  if (match.first_goal_minute == null) {
    return true;
  }
  return total > coveredTotal;
}

/** Display string for a live minute including added time, e.g. "45+2" or "67". */
export function formatMinuteDisplay(minute?: number | null, extra?: number | null): string | null {
  if (minute == null) {
    return null;
  }
  return extra && extra > 0 ? `${minute}+${extra}` : `${minute}`;
}

/** First real minute (≤30) the favorite scored, when the timeline is known. */
export function favoriteEarlyGoalMinute(match: FlashscoreMatch): number | null {
  if (match.favorite_side !== "home" && match.favorite_side !== "away") {
    return null;
  }
  const minutes = match.favorite_side === "away" ? match.away_goal_minutes : match.home_goal_minutes;
  const early = (minutes ?? []).filter((minute) => minute <= EARLY_GOAL_MINUTE);
  return early.length ? Math.min(...early) : null;
}

function uniqueSortedMinutes(minutes: number[]): number[] {
  return [...new Set(minutes)].sort((left, right) => left - right);
}

export function withEarlyGoalFlags(match: FlashscoreMatch): FlashscoreMatch {
  const minute = match.minute;
  const homeScore = match.home_score ?? 0;
  const awayScore = match.away_score ?? 0;
  const totalGoals = homeScore + awayScore;
  const favoriteScore = match.favorite_side === "away" ? awayScore : homeScore;
  const inEarlyWindow = minute != null && minute <= EARLY_GOAL_MINUTE;

  // Real goal minutes from the SofaScore timeline take priority over the poll-minute guess.
  const homeMinutes = uniqueSortedMinutes(match.home_goal_minutes ?? []);
  const awayMinutes = uniqueSortedMinutes(match.away_goal_minutes ?? []);
  const favoriteMinutes = match.favorite_side === "away" ? awayMinutes : homeMinutes;
  const earlyIncidentMinutes = [...homeMinutes, ...awayMinutes].filter((value) => value <= EARLY_GOAL_MINUTE);
  const favoriteEarlyIncident = favoriteMinutes.filter((value) => value <= EARLY_GOAL_MINUTE);
  const hasIncidents = homeMinutes.length > 0 || awayMinutes.length > 0;

  const favoriteWatched =
    match.favorite_team != null &&
    match.favorite_odds != null &&
    match.favorite_odds <= ALERT_ODDS_THRESHOLD;

  const sawEarlyGoal = hasIncidents
    ? Boolean(match.early_goal) || earlyIncidentMinutes.length > 0
    : Boolean(match.early_goal) || (inEarlyWindow && totalGoals > 0);
  const sawEarlyFavoriteGoal = hasIncidents
    ? Boolean(match.early_favorite_goal) || (favoriteWatched && favoriteEarlyIncident.length > 0)
    : Boolean(match.early_favorite_goal) || (inEarlyWindow && favoriteWatched && favoriteScore > 0);

  let earlyGoalMinute: number | null;
  if (earlyIncidentMinutes.length > 0) {
    earlyGoalMinute = Math.min(...earlyIncidentMinutes);
  } else if (match.early_goal_minute != null) {
    earlyGoalMinute = match.early_goal_minute;
  } else if (sawEarlyGoal && inEarlyWindow) {
    earlyGoalMinute = minute ?? null;
  } else {
    earlyGoalMinute = null;
  }

  // First goal of the match (any team), taken only from the timeline and kept sticky
  // once detected. Never inferred from the current scoreline.
  const allIncidentMinutes = [...homeMinutes, ...awayMinutes].sort((left, right) => left - right);
  const firstGoalMinute = allIncidentMinutes.length > 0 ? allIncidentMinutes[0] : match.first_goal_minute ?? null;
  const goalUnder30 = firstGoalMinute != null && firstGoalMinute <= EARLY_GOAL_MINUTE;

  return {
    ...match,
    home_goal_minutes: homeMinutes,
    away_goal_minutes: awayMinutes,
    early_goal: sawEarlyGoal,
    early_favorite_goal: sawEarlyFavoriteGoal,
    early_goal_minute: earlyGoalMinute,
    first_goal_minute: firstGoalMinute,
    goal_under_30: goalUnder30,
    alert_eligible: sawEarlyFavoriteGoal || isAlertEligible(match),
  };
}

export function isAlertEligible(match: FlashscoreMatch) {
  if (match.early_favorite_goal) {
    return true;
  }
  if (
    !match.favorite_team ||
    match.favorite_odds == null ||
    match.favorite_odds > ALERT_ODDS_THRESHOLD ||
    match.minute == null ||
    match.minute > EARLY_GOAL_MINUTE ||
    match.home_score == null ||
    match.away_score == null
  ) {
    return false;
  }
  const favoriteScore = match.favorite_side === "away" ? match.away_score : match.home_score;
  return (favoriteScore || 0) > 0;
}

export function sortFlashscoreMatches(matches: FlashscoreMatch[]) {
  return [...matches].sort((left, right) => {
    const leftRank = earlyGoalRank(left);
    const rightRank = earlyGoalRank(right);
    if (leftRank !== rightRank) {
      return leftRank - rightRank;
    }
    const leftStart = left.start_time || "";
    const rightStart = right.start_time || "";
    return leftStart.localeCompare(rightStart);
  });
}

/** Faster refresh while a ≤1.50 favorite can still trigger the early-goal signal. */
export function liveRefreshIntervalMs(matches: FlashscoreMatch[], now = Date.now()): number {
  return matches.some((match) => isCriticalSignalWatch(match, now))
    ? FAST_LIVE_REFRESH_MS
    : SLOW_LIVE_REFRESH_MS;
}

export function isCriticalSignalWatch(match: FlashscoreMatch, now = Date.now()): boolean {
  if (match.favorite_odds == null || match.favorite_odds > ALERT_ODDS_THRESHOLD) {
    return false;
  }
  if (match.early_favorite_goal || match.alert_eligible) {
    return false;
  }
  const status = (match.status || "").toLowerCase();
  if (status.includes("finish") || status.includes("ended") || status.includes("afterpen")) {
    return false;
  }
  if (match.minute != null) {
    return match.minute <= EARLY_GOAL_MINUTE + 10;
  }
  if (!match.start_time) {
    return false;
  }
  const start = new Date(match.start_time).getTime();
  if (!Number.isFinite(start)) {
    return false;
  }
  // Kickoff window: 20 minutes before to 50 minutes after scheduled start.
  return start <= now + 20 * 60_000 && start >= now - 50 * 60_000;
}

function earlyGoalRank(match: FlashscoreMatch) {
  if (match.early_favorite_goal || match.alert_eligible) return 0;
  if (match.early_goal) return 1;
  if (match.minute != null && match.minute <= EARLY_GOAL_MINUTE) return 2;
  return 3;
}

function isFlashscoreMatch(value: unknown): value is FlashscoreMatch {
  if (!value || typeof value !== "object") {
    return false;
  }
  const match = value as FlashscoreMatch;
  return typeof match.event_id === "string" &&
    typeof match.home_team === "string" &&
    typeof match.away_team === "string" &&
    typeof match.competition === "string";
}
